s3_artifact/action.py

```diff
@@ -0,0 +1,101 @@
+import os
+import subprocess
+from dataclasses import dataclass
+from distutils.util import strtobool
+from pathlib import Path
+from typing import Optional, Sequence
+
+from dataclasses_json import dataclass_json
+from yaml import safe_load
+
+DEFAULT_MIME_TYPES = {
+    ".css": "text/css",
+    ".html": "text/html",
+    ".js": "application/javascript",
+    ".jpg": "image/jpeg",
+    ".jpeg": "image/jpeg",
+    ".pdf": "application/pdf",
+    ".png": "image/png",
+}
+
+S3_NO_PROGRESS = "--no-progress"
+S3_CP_OPTIONS = f"--recursive {S3_NO_PROGRESS}"
+S3_SYNC_OPTIONS = f"--delete {S3_NO_PROGRESS}"
+
+
+@dataclass_json
+@dataclass
+class WebsiteCustomMetadataConfig:
+    path: str
+    cache_control: str
+    mime_type: Optional[str] = None
+
+
+@dataclass_json
+@dataclass
+class WebsiteConfig:
+    target_buckets: dict[str, str]
+    local_artifacts_path: str
+    default_cache_control: str
+    custom_metadata: Optional[tuple[WebsiteCustomMetadataConfig, ...]] = ()
+
+
+def _run_commands(commands: Sequence[str], dry_run: bool):
+    for command in commands:
+        print(command)
+
+        if not dry_run:
+            subprocess.run(command, shell=True, check=True)
+
+
+def deploy(config: WebsiteConfig, artifacts_s3_path: str, environment: str) -> Sequence[str]:
+    target = f"s3://{config.target_buckets[environment]}"
+    return (
+        f"aws s3 ls {artifacts_s3_path}",  # Check if artifact exists, otherwise fail deployment
+        # cp and sync because there are problems syncing all files - https://github.com/aws/aws-cli/issues/3273
+        f"aws s3 cp {artifacts_s3_path} {target} {S3_CP_OPTIONS}",
+        f"aws s3 sync {artifacts_s3_path} {target} {S3_SYNC_OPTIONS}",
+    )
+
+
+def upload(config: WebsiteConfig, target: str) -> Sequence[str]:
+    def _prepare_metadata_command(cache_config: WebsiteCustomMetadataConfig):
+        content_type = cache_config.mime_type or DEFAULT_MIME_TYPES[Path(cache_config.path).suffix]
+        content_type_option = f"--content-type '{content_type}'"
+
+        return (
+            f"aws s3 cp {target} {target} {S3_CP_OPTIONS} --exclude '*' --include {cache_config.path} "
```

Comment:
I was wondering how the `--recursive` option and the `**` glob patterns interact, but I guess it works somehow :)